src/conversations/Conversations.ts

```diff
@@ -8,49 +9,108 @@ import { buildUserIntroTopic, buildUserInviteTopic } from '../utils'
 import { messageApi, fetcher } from '@xmtp/proto'
+import { SortDirection } from '../ApiClient'
 const { b64Decode } = fetcher
 
 const messageHasHeaders = (msg: MessageV1): boolean => {
   return Boolean(msg.recipientAddress && msg.senderAddress)
 }
+
+class ConversationCache {
+  private conversations: Conversation[]
+  private mutex: Mutex
+  private latestSeen?: Date
+  private seenTopics: Set<string>
+
+  constructor() {
+    this.conversations = []
+    this.mutex = new Mutex()
+    this.seenTopics = new Set()
+  }
+
+  async getAndLock() {
+    const release = await this.mutex.acquire()
+    return {
+      // Release MUST be called. All callers should use a try/finally block
+      release,
+      latestSeen: this.latestSeen,
+      // addConvos adds new conversations to the cache and returns the complete list
+      addConvos: (newConvos: Conversation[]): Conversation[] => {
+        for (const convo of newConvos) {
+          if (!this.seenTopics.has(convo.topic)) {
+            this.seenTopics.add(convo.topic)
+            this.conversations.push(convo)
+            if (!this.latestSeen || convo.createdAt > this.latestSeen) {
+              this.latestSeen = convo.createdAt
+            }
+          }
+        }
+        // Create a new array so that callers can't accidentally modify the cache
+        return [...this.conversations]
+      },
+    }
+  }
+}
+
 /**
  * Conversations allows you to view ongoing 1:1 messaging sessions with another wallet
  */
 export default class Conversations {
   private client: Client
+  private v2Cache: ConversationCache
+
   constructor(client: Client) {
     this.client = client
+    this.v2Cache = new ConversationCache()
   }
 
   /**
    * List all conversations with the current wallet found in the network, deduped by peer address
    */
   async list(): Promise<Conversation[]> {
-    const [seenPeers, invitations] = await Promise.all([
+    const [seenPeers, v2Convos] = await Promise.all([
       this.getIntroductionPeers(),
-      this.client.listInvitations(),
+      this.listV2Conversations(),
     ])
 
-    const conversations: Conversation[] = []
+    const v1Convos: Conversation[] = []
     seenPeers.forEach((sent, peerAddress) =>
-      conversations.push(new ConversationV1(this.client, peerAddress, sent))
+      v1Convos.push(new ConversationV1(this.client, peerAddress, sent))
     )
 
-    for (const sealed of invitations) {
-      try {
-        const unsealed = await sealed.v1.getInvitation(this.client.keys)
-        conversations.push(
-          await ConversationV2.create(this.client, unsealed, sealed.v1.header)
-        )
-      } catch (e) {
-        console.warn('Error decrypting invitation', e)
-      }
-    }
+    const conversations = v1Convos.concat(v2Convos)
 
     conversations.sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime())
     return conversations
   }
 
+  private async listV2Conversations(): Promise<Conversation[]> {
```

Comment:
I think this works as expected. Wondering why not hide this inside the cache to avoid potential mistake of forgetting to release the lock. Not feeling strongly about this, feel free to keep as is.

Reply:
I considered having the cache take a callback function to execute so that it could handle the releasing itself. I'm still playing around with the ergonomics and might do that.

I do want this to be extensible enough that I can do the same thing for V1 conversations using the same cache class, which that would allow

Reply:
I updated to use a callback and I think that's a lot tidier. Makes it impossible for callers to leave the cache locked.